promotion/drain.go

```diff
@@ -542,51 +543,52 @@ func (service *Service) RedeemAndTransferFunds(ctx context.Context, credentials
 	}
 
 	if ok, _ := appctx.GetBoolFromContext(ctx, appctx.ReputationOnDrainCTXKey); ok {
-		// perform reputation check for wallet, and error accordingly if there is a reputation failure
-		reputable, err := service.reputationClient.IsWalletAdsReputable(ctx, walletID, "")
-		if err != nil {
-			logger.Error().Err(err).Msg("RedeemAndTransferFunds: failed to check reputation of wallet")
-			return nil, errReputationServiceFailure
-		}
-
-		if !reputable {
-			return nil, errWalletNotReputable
-		}
-	}
+		// are we running in withdrawal limit mode?
+		if ok, _ := appctx.GetBoolFromContext(ctx, appctx.ReputationWithdrawalOnDrainCTXKey); ok {
+			// tally up all prior claims on this promotion for all linked provider accounts associated
+			// as the "withdrawalAmount"
+			promotionID, withdrawalAmount, err := service.Datastore.GetWithdrawalsAssociated(&walletID, claimID)
+			if err != nil {
+				logger.Error().Err(err).Msg("RedeemAndTransferFunds: failed to lookup associated withdrawals")
+				return nil, fmt.Errorf("failed to lookup associated withdrawals: %w", err)
+			}
 
-	/* TODO: for nitro enablement we will use the prepare api to send individual txs on a batch
+			if promotionID == nil {
+				logger.Error().Err(err).Msg("RedeemAndTransferFunds: failed to lookup associated withdrawals")
+				return nil, fmt.Errorf("failed to lookup associated withdrawals: no matching promotion")
+			}
 
-	// use paymentsClient to "prepare" transfer with batch id
-	resp, err := service.paymentsClient.Prepare(ctx, &paymentspb.PrepareRequest{
-		BatchMeta: &paymentspb.BatchMeta{
-			// TODO: get the batch id in here
-		},
-		State:     paymentspb.State_PREPARED,
-		Custodian: paymentspb.Custodian(paymentspb.Custodian_value[strings.ToUpper(*wallet.UserDepositAccountProvider)]),
-		BatchTxs: []*pb.Transaction{
-			{
-				Destination: wallet.UserDepositDestination,
-				Amount:      total.String(),
-				Currency:    "BAT",
-			},
-		},
-	})
-	if err != nil {
-		logger.Error().Err(err).Msg("failed to make connection to payments")
-		return nil, fmt.Errorf("failed to call prepare for payments transfer: %w", err)
-	}
+			// perform reputation check for wallet, and error accordingly if there is a reputation failure
+			reputable, justification, err := service.reputationClient.IsDrainReputable(ctx, walletID, *promotionID, withdrawalAmount)
+			if err != nil {
+				logger.Error().Err(err).Msg("RedeemAndTransferFunds: failed to check reputation of wallet")
+				return nil, errReputationServiceFailure
+			}
 
-	tx := new(walletutils.TransactionInfo)
+			if !reputable {
+				// pass along the justification in the response error
+				switch justification {
+				case "exceeded_withdrawal_limit":
```

Comment:
This justification won't match with the response from reputation in case of exceeded withdrawal amount. I'd suggest you parsing the cohort here rather than the justification. And use the justification only for human readability purposes (DB/logs/errors etc)

Reply:
addressed here: https://github.com/brave-intl/bat-go/pull/1263 @gpestana can you +1 there?